include/depthimage_to_laserscan/DepthImageToLaserScan.h

```diff
@@ -181,31 +187,81 @@ namespace depthimage_to_laserscan
+      } catch (tf::TransformException &ex){
+        ROS_ERROR("%s", ex.what());
+        return;
+      }
+
+      // store specific element we want to use
+      double tf_basis_2_0 = transform.getBasis()[2][0];
+      double tf_basis_2_1 = transform.getBasis()[2][1];
+      double tf_basis_2_2 = transform.getBasis()[2][2];
+      double tf_origin_z = transform.getOrigin().z();
+
       for(int v = offset; v < offset+scan_height_; v++, depth_row += row_step){
-		for (int u = 0; u < (int)depth_msg->width; u++) // Loop over each pixel in row
-		{	
-		  T depth = depth_row[u];
-		  
-		  double r = depth; // Assign to pass through NaNs and Infs
-		  double th = -atan2((double)(u - center_x) * constant_x, unit_scaling); // Atan2(x, z), but depth divides out
-		  int index = (th - scan_msg->angle_min) / scan_msg->angle_increment;
+	for (int u = 0; u < (int)depth_msg->width; u++) // Loop over each pixel in row
+	{	
+	  T depth = depth_row[u];
 		  
-		  if (depthimage_to_laserscan::DepthTraits<T>::valid(depth)){ // Not NaN or Inf
-		    // Calculate in XYZ
-		    double x = (u - center_x) * depth * constant_x;
-		    double z = depthimage_to_laserscan::DepthTraits<T>::toMeters(depth);
-		    
-		    // Calculate actual distance
-		    r = sqrt(pow(x, 2.0) + pow(z, 2.0));
-		  }
+	  double r = depth; // Assign to pass through NaNs and Infs
+          // Because the camera is flipped, the theta should also be flipped
+	  double th = atan2((double)(u - center_x) * constant_x, unit_scaling); // Atan2(x, z), but depth divides out
+	  int index = (th - scan_msg->angle_min) / scan_msg->angle_increment;
+
+	  if (depthimage_to_laserscan::DepthTraits<T>::valid(depth)){ // Not NaN or Inf
+	    // Calculate in XYZ
+            double x = (u - center_x) * depth * constant_x;
+            double y = (v - center_y) * depth * constant_y;
+	    double z = depthimage_to_laserscan::DepthTraits<T>::toMeters(depth);
+
+            // Early return while z is out of range
+            if(z > range_max_ || z < range_min_){
+              continue;
+            }
+
+            double rectified_height = x * tf_basis_2_0 + y * tf_basis_2_1 + z * tf_basis_2_2 + tf_origin_z;
+            //double rectify_height = transform.getBasis()[2][0] * x + transform.getBasis()[2][1] * y + transform.getBasis()[2][2] * z + transform.getOrigin().z();
+
+            /*
+            geometry_msgs::PointStamped cur_point;
+            geometry_msgs::PointStamped trans_point;
+            cur_point.header.frame_id = "camera_depth_optical_frame";
+            cur_point.header.stamp= ros::Time();
+            cur_point.point.x = x;
+            cur_point.point.y = y;
+            cur_point.point.z = z;
+
+            try{
+              listener_.transformPoint("/map", cur_point, trans_point);
+            } catch(tf::TransformException &ex) {
+              ROS_ERROR("%s", ex.what());
+              ros::Duration(1.0).sleep();
+              continue;
```

Comment:
This is the part I was not sure it's a good idea and thus left the commented code there, will clean it up before merging.